libdevice/sanitizer_utils.cpp

```diff
@@ -286,8 +286,17 @@ bool MemIsZero(__SYCL_GLOBAL__ const char *beg, uptr size) {
 bool __asan_internal_report_save(DeviceSanitizerErrorType error_type) {
   const int Expected = ASAN_REPORT_NONE;
   int Desired = ASAN_REPORT_START;
-  auto &SanitizerReport =
-      ((__SYCL_GLOBAL__ LaunchInfo *)__AsanLaunchInfo)->SanitizerReport;
+
+  // work-group linear id
+  const auto WG_LID =
+      __spirv_BuiltInWorkgroupId.x * __spirv_BuiltInNumWorkgroups.y *
+          __spirv_BuiltInNumWorkgroups.z +
+      __spirv_BuiltInWorkgroupId.y * __spirv_BuiltInNumWorkgroups.z +
+      __spirv_BuiltInWorkgroupId.z;
+
+  auto &SanitizerReport = ((__SYCL_GLOBAL__ LaunchInfo *)__AsanLaunchInfo)
+                              ->SanitizerReport[WG_LID % ASAN_MAX_NUM_REPORTS];
```

Comment:
Given an error in a workgroup and another error in an another workgroup, these two errors may be saved to the same index and either of them has a chance to be kept.
Program behavior is probably more predicable if we use an atomically increasing index instead.

Reply:
> Program behavior is probably more predicable if we use an atomically increasing index instead.

What does predicable mean here? 
Even using atomic counter, we can't guarantee that we can save all error reports in your case, and their order are also undetermined.
I didn't try to save as most as possible error reports because I think it's not necessary (as most of them are likely pointing to the same error location).

Maybe it would be better to save more different types of error, but it's too complicated to implement now.

Reply:
> Even using atomic counter, we can't guarantee that we can save all error reports in your case, and their order are also undetermined.

Right, order is indeterministic. Could you please elaborate why we can't guarantee that all of the two errors are not reported in my case?

Reply:
> Given an error in a workgroup and another error in an another workgroup, these two errors may be saved to the same index and either of them has a chance to be kept.

I don't understand, why does either of them has a chance to be kept?
They can save into specific index of reports unless that index is already used.

Reply:
Because the order of workgroup execution is indeterministic.
When there is clash in indexing, we can't say which one is saved since the second error won't be save after the first one is saved.

Reply:
> Because the order of workgroups execution is indeterministic.
> When there is clash in indexing, we can't say which one is saved since the second error won't be save after the first one is saved.

Yep. I think it's okay to save either one.